test/test_einsum.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import absolute_import, division, print_function
+
+import itertools
+import pytest
+from collections import OrderedDict
+
+import torch
+from pyro.ops.contract import naive_ubersum
+
+import funsor
+
+
+def make_example(equation, fill=None, sizes=(2, 3)):
+    symbols = sorted(set(equation) - set(',->'))
+    sizes = {dim: size for dim, size in zip(symbols, itertools.cycle(sizes))}
+    inputs, outputs = equation.split('->')
+    inputs = inputs.split(',')
+    outputs = outputs.split(',')
+    operands = []
+    for dims in inputs:
+        shape = tuple(sizes[dim] for dim in dims)
+        operands.append(torch.randn(shape) if fill is None else torch.full(shape, fill))
+    return inputs, outputs, operands, sizes
+
+
+def naive_einsum(eqn, *terms):
+    assert isinstance(eqn, str)
+    assert all(isinstance(term, funsor.Funsor) for term in terms)
+    inputs, output = eqn.split('->')
+    input_dims = frozenset(d for inp in inputs.split(',') for d in inp)
+    output_dims = frozenset(d for d in output)
+    reduce_dims = tuple(d for d in input_dims - output_dims)
+    prod = terms[0]
+    for term in terms[1:]:
+        prod = prod * term
+    for reduce_dim in reduce_dims:
+        prod = prod.sum(reduce_dim)
+    return prod
+
+
+def naive_plated_einsum(eqn, *terms, plates='', modulo_total=False):
+    assert isinstance(eqn, str)
+    assert all(isinstance(term, funsor.Funsor) for term in terms)
+    # ...
+    raise NotImplementedError("TODO implement naive plated einsum")
+
+
+EINSUM_EXAMPLES = [
+    "a,b->",
+    "ab,a->",
+    "a,a->",
+    "a,a->a",
+    "a,a,a,ab->ab",
```

Comment:
We should add an xfailing example
```
"ab->ba"
```
To make this pass we should resurrect [.align()](https://github.com/pyro-ppl/funsor/blob/da322019d355bd43d564a6b5b42c65402b2df064/funsor/terms.py#L105) and [Align](https://github.com/pyro-ppl/funsor/blob/da322019d355bd43d564a6b5b42c65402b2df064/funsor/terms.py#L388). I'll add this as a separate issue.

Reply:
I've added issue #26